### DISCUSSION TITLE
How can I call a function with Ref parameter function with WidgetRef?

### DISCUSSION THREAD

**Discussion by vishnuagbly** [rrousselGit/riverpod#1651] [Q&A]
I was creating an application where, I need to call a function, which will take Ref as input (as it can be called by the providers), but I also need to call it using WidgetRef, while listening to a StateNotifier Provider.

```dart
final firstProvider = ... //Initialization

final someProvider = Provider<String>((ref) {
  performOperation(ref);
  return '';
});

class MyWidget extends ConsumerWidget {
...

  Widget build(BuildContext context, WidgetRef ref) {
    ref.listen(firstProvider, (pref, next) => performOperation(ref));
    return ...;
  }
}
```

Where the `performOperation` function might be like this:-
```dart
void performOperation(Ref ref) {
  ref.read(...);
  ref.read(...);
  performNextOperation(ref);
  ...
}
```

I do not want to pass the objects read by providers directly to the function, as the number of parameters might change with time, and also there are many more functions, as static members of different Classes which are called one after another, taking Ref as the argument.

**Comment by rrousselGit** [comment]
That's the thing: you don't.
Get a real Ref instead.


In your case, you probably want to make a StateNotiier, and have `performOperation` be a method of that StateNotifier. Then you'd have a Ref from the StateNotifierProvider

**Reply by vishnuagbly** [reply]
I was wondering if I can make the `performOperation` as a provider like this:-
```dart
final performOperation = Provider<void>((ref) {
  ref.read(...);
  ref.read(...);
  performNextOperation(ref);
  ...
});
```

then call it by reading it like:-
```dart
final firstProvider = ... //Initialization

final someProvider = Provider<String>((ref) {
  ref.read(performOperation);
  return '';
});

class MyWidget extends ConsumerWidget {
...

  Widget build(BuildContext context, WidgetRef ref) {
    ref.listen(firstProvider, (pref, next) => ref.read(performOperation));
    return ...;
  }
}
```

@rrousselGit  Would this approach work? Are there any problems that might happen with this?

**Reply by rrousselGit** [reply]
It's just a weird roundabout way of doing it.

You should be using a StateNotifier/ChangeNotifier/WhateverNotifier.

**Reply by vishnuagbly** [reply]
So, should the `performOperation` be like this:-
```dart
class SomeNotifier extends StateNotifier<void> {
  final Ref _ref;

  SomeNotifier(this._ref): super(null);
  
  void performOperation() {
    _ref.read(...);
    _ref.read(...);
    performNextOperation(_ref);
    ...
  }
}

final someProvider = StateNotifierProvider<SomeNotifier, void>((ref) {
  return SomeNotifier(ref);
});
```

then call it by reading it like:-
```dart
final firstProvider = ... //Initialization

final someProvider = Provider<String>((ref) {
  ref.read(someProvider.notifier).performOperation();
  return '';
});

class MyWidget extends ConsumerWidget {
...

  Widget build(BuildContext context, WidgetRef ref) {
    ref.listen(firstProvider, (pref, next) => ref.read(someProvider.notifier).performOperation());
    return ...;
  }
}
```

@rrousselGit Is this what you might be saying? It just looks like the similar thing, but with more boilerplate 😅.

**Reply by rrousselGit** [reply]
No. I meant about changing `someProvider` to a `StateNotifierProvider`

**Reply by vishnuagbly** [reply]
@rrousselGit, Sorry, I named two providers the same in the above reply, I meant something like this in the above reply:-
```dart
class OperationNotifier extends StateNotifier<void> {
  final Ref _ref;

  OperationNotifier(this._ref): super(null);
  
  void performOperation() {
    _ref.read(...);
    _ref.read(...);
    performNextOperation(_ref);
    ...
  }
}

final operationProvider = StateNotifierProvider<SomeNotifier, void>((ref) {
  return OperationNotifier(ref);
});
```

then call it by reading it like:-
```dart
final firstProvider = ... //Initialization

final someProvider = Provider<String>((ref) {
  ref.read(operationProvider.notifier).performOperation();
  return '';
});

class MyWidget extends ConsumerWidget {
...

  Widget build(BuildContext context, WidgetRef ref) {
    ref.listen(firstProvider, (pref, next) => ref.read(operationProvider.notifier).performOperation());
    return ...;
  }
}
```

**Reply by davidmartos96** [reply]
@rrousselGit What would be the benefit of using a Notifier like class in this example? It doesn't hold any state.
Wouldn't it be enough to use a regular provider here?

**Reply by rrousselGit** [reply]
That's not what I said. Don't do `STateNotifier<void>`

It's about changing your existing `Provider<String>`, not creating a new provider

**Reply by vishnuagbly** [reply]
@rrousselGit But what is the advantage of that, I still need to call `performOperation` from the `build` method, where I only have access to `WidgetRef`, and, `someProvider` basically just works like a simple function, it does not need to manage state, better way to represent `someProvider` in my above examples would be like this:-
```dart
final someProvider = Provider<String>((ref) {
  performOperation(ref);
  var value = ref.read(...
  ... //reading some other providers, performing some simple transformation to the 'value' 
  return value;
});
```

**Reply by rrousselGit** [reply]
If you feel like "someProvider" isn't the correct one, try "firstProvider" then
But the logic is the same. You should refactor one of your providers to a StateNotifier. That's how we trigger side-effects in Riverpod


As for your latest snippet, you shouldn't perform side-effects within the initialization of a provider. So you can't do that

**Reply by vishnuagbly** [reply]
@rrousselGit In my actual case, `firstProvider` is already a `StateNotifierProvider` 😂.
See the thing is, like this, let's assume I want to update the state of a few providers, let's assume `performOperation` performs some checks by reading a few providers and updates the value of a `countProvider` accordingly and I want to update it's value whenever the `someProvider` is executed, to generate a string. And also whenever the state of `firstProvider` (StateNotifierProvider) changes.
Therefore my initial problem will become like this:-
```dart
final firstProvider = ... //Initialization

final someProvider = Provider<String>((ref) {
  performOperation(ref);
  var value = ref.read(...
  ... //reading some other providers, performing some simple transformation to the 'value' 
  return value;
});

class MyWidget extends ConsumerWidget {
...

  Widget build(BuildContext context, WidgetRef ref) {
    ref.listen(firstProvider, (pref, next) => performOperation(ref));
    return ...;
  }
}
```
And my `performOperation` function looks like this:-
```dart
void performOperation(Ref ref) {
  performChecks(ref); //Chain of functions performing some checks, by reading ultiple providers.
  final value = getIncrementValue(ref); //Reading multiple providers to generate the value.
  ref.read(countProvider.notifier).increment(value);
}
```

Also if possible can you provide a pseudo-code or snippet for your answer?
